vhost-device-gpu/src/backend/virgl.rs

```diff
@@ -159,7 +161,17 @@ impl VirglRendererAdapter {
             fence_state.clone(),
         ));
 
-        let renderer = VirglRenderer::init(virglrenderer_flags, fence_handler, None)
+        // Open the GPU device if a path was specified
+        let render_server_fd = config.flags().gpu_path.as_ref().map(|path| {
+            File::open(path)
+                .unwrap_or_else(|e| {
+                    error!("Failed to open GPU device at {}: {}", path.display(), e);
+                    exit(1);
+                })
```

Comment:
calling `exit(1)` here is not a good idea, not different from `panic()` (I'll say it's even worse, because `panic()` can give you some information), I think you should change `new() -> Result<Self, ...>` and return an error, also leaving the caller to call `error!()`, so this could be just : `File::open(path).map_err(|e| ...)?`

Reply:
IMO, `exit(1)` is actually reasonable for initialization/startup errors like an invalid GPU path, Initially I considered propagating the error, just thought it would make it more complex having to propagate the error up the call stack.

Reply:
Virglrenderer can receive a `-1` as the render server fd. In this case, virglrenderer will find an available GPU by itself. Can we print some warnings to inform users that they passed an invalid GPU path, but not block vhost-device-gpu from starting?

Reply:
Even if it's harmless to `exit` now, it might not be with future changes. Destructors/cleanups won't run if you `exit`. Returning an error is trivial

Reply:
Thanks for the feedback all, I refactored the code to return `io::Result<Self>` . 
For the GPU path, if the file can't be opened let's say a typo in the path, it logs a warning and falls back to None which translates to -1 for the render server fd, hence virglrenderer can auto-select a GPU. But new() can still return errors for other initialization failures.